Mathlib/Algebra/Hom/Equiv/Basic.lean

```diff
@@ -0,0 +1,659 @@
+  EquivLike.bijective e
+#align mul_equiv.bijective MulEquiv.bijective
+
+@[to_additive]
+protected theorem injective (e : M ≃* N) : Function.Injective e :=
+  EquivLike.injective e
+#align mul_equiv.injective MulEquiv.injective
+
+@[to_additive]
+protected theorem surjective (e : M ≃* N) : Function.Surjective e :=
+  EquivLike.surjective e
+#align mul_equiv.surjective MulEquiv.surjective
+
+/-- The identity map is a multiplicative isomorphism. -/
+@[refl, to_additive "The identity map is an additive isomorphism."]
+def refl (M : Type _) [Mul M] : M ≃* M :=
+  { Equiv.refl _ with map_mul' := fun _ _ => rfl }
+#align mul_equiv.refl MulEquiv.refl
+
+@[to_additive]
+instance : Inhabited (M ≃* M) :=
+  ⟨refl M⟩
+
+/-- The inverse of an isomorphism is an isomorphism. -/
+@[symm, to_additive "The inverse of an isomorphism is an isomorphism."]
+def symm (h : M ≃* N) : N ≃* M :=
+  { h.toEquiv.symm with map_mul' := (h.toMulHom.inverse h.toEquiv.symm h.left_inv h.right_inv).map_mul }
+#align mul_equiv.symm MulEquiv.symm
+
+@[simp, to_additive]
+theorem inv_fun_eq_symm {f : M ≃* N} : f.invFun = f.symm :=
+  rfl
+#align mul_equiv.inv_fun_eq_symm MulEquiv.inv_fun_eq_symm
+
+-- we don't hyperlink the note in the additive version, since that breaks syntax highlighting
+-- in the whole file.
+/-- See Note [custom simps projection] -/
+@[to_additive "See Note custom simps projection"]
+def Simps.symmApply (e : M ≃* N) : N → M :=
+  e.symm
+#align mul_equiv.simps.symm_apply MulEquiv.Simps.symmApply
+
+initialize_simps_projections AddEquiv (toFun → apply, invFun → symmApply)
+
+initialize_simps_projections MulEquiv (toFun → apply, invFun → symmApply)
+
+@[simp, to_additive]
+theorem to_equiv_symm (f : M ≃* N) : f.symm.toEquiv = f.toEquiv.symm :=
+  rfl
+#align mul_equiv.to_equiv_symm MulEquiv.to_equiv_symm
+
+@[simp, to_additive]
+theorem coe_mk (f : M → N) (g h₁ h₂ h₃) : ⇑(MulEquiv.mk f g h₁ h₂ h₃) = f :=
+  rfl
+#align mul_equiv.coe_mk MulEquiv.coe_mk
+
+@[simp, to_additive]
+theorem to_equiv_mk (f : M → N) (g : N → M) (h₁ h₂ h₃) : (mk f g h₁ h₂ h₃).toEquiv = ⟨f, g, h₁, h₂⟩ :=
+  rfl
+#align mul_equiv.to_equiv_mk MulEquiv.to_equiv_mk
+
+@[simp, to_additive]
+theorem symm_symm : ∀ f : M ≃* N, f.symm.symm = f
+  | ⟨f, g, h₁, h₂, h₃⟩ => rfl
+#align mul_equiv.symm_symm MulEquiv.symm_symm
+
+@[to_additive]
+theorem symm_bijective : Function.Bijective (symm : M ≃* N → N ≃* M) :=
+  Equiv.bijective ⟨symm, symm, symm_symm, symm_symm⟩
+#align mul_equiv.symm_bijective MulEquiv.symm_bijective
+
+@[simp, to_additive]
+theorem symm_mk (f : M → N) (g h₁ h₂ h₃) :
+    (MulEquiv.mk f g h₁ h₂ h₃).symm = { (MulEquiv.mk f g h₁ h₂ h₃).symm with toFun := g, invFun := f } :=
+  rfl
+#align mul_equiv.symm_mk MulEquiv.symm_mk
+
+@[simp, to_additive]
+theorem refl_symm : (refl M).symm = refl M :=
+  rfl
+#align mul_equiv.refl_symm MulEquiv.refl_symm
+
+/-- Transitivity of multiplication-preserving isomorphisms -/
+@[trans, to_additive "Transitivity of addition-preserving isomorphisms"]
```

Comment:
Bug with `trans`, similar to `symm` [#855](https://github.com/leanprover-community/mathlib4/pull/855)